### DISCUSSION TITLE
Serialize/Deserialize json on non-blocking file descriptor

### DISCUSSION THREAD

**Discussion by milasudril** [nlohmann/json#4017] [Q&A]
Is it possible to serialize/deserialize json when using non-blocking I/O?

For parsing, I imagine a `parse_context`, that contains the current state, and a function such as

```c++
void parse_ctxt::parse(std::span<char const>);
```

If the last `}` (or `]` if the top-level obejct is an array) has been reached, The `parse_context` should contain a valid json object. Otherwise, it should contain a flag that indicates the status. The status could also indicate some error condition.

Example usage:

```c++
class json_reader
{
public:
  enum class parse_result{keep_going, completed};
  
  auto read_data()
  {
    std::array<char, 4096> buffer{};
    while(!ctxt.completed()) {
      auto res = read(fd, std::data(buffer), std::size(buffer));
      if(res == -1 && errno == EAGAIN)
      { return keep_going; }
      
      ctxt.parse(std::span{std::data(buffer), static_cast<size_t>(res)});
    }
    return completed;
  }
  
  auto take_result()
  { return std::move(ctxt.result()); }

private:
  int fd;
  parse_context ctxt;
};
```
The `parse_context` may also contain a value that indicates where in the buffer, the json object ended. This provides the options to

1. Continue to parse a different json object
2. Continue to read some other form of data
3. Treat it as an error condition if the block passed to `parse` was not entirely consumed

**Comment by nlohmann** [comment]
The library doesn't currently support suspending and resuming a partial parse across separate, non-contiguous read() calls the way you describe. The SAX interface (json::sax_parse, docs/mkdocs/docs/features/parsing/sax_interface.md) still needs to be driven by a single input source (a string, iterator range, or istream) that it consumes in one call, and if that source blocks/returns EAGAIN mid-read there's no supported way to pause the internal lexer state and feed it more bytes later. The practical workaround today is to keep accumulating bytes from the non-blocking fd into your own buffer (e.g. std::string) until you can determine (via your own framing, e.g. counting braces/brackets or a length prefix) that a complete JSON document is present, then call json::parse() once on the complete buffer; there's no lower-level resumable parse_context exposed publicly.

---
*This reply was drafted by [Claude Code](https://claude.com/claude-code) on behalf of @nlohmann, as part of a review of open discussions.*
